File: Dynamic Programming/11053. 가장 긴 증가하는 부분 수열/11053.java
import java.io.*;
import java.util.*;

public class Main {
    public static final int MAX = 1000;
    public static int[] d = new int[MAX + 1];

    public static int lis_dp(int n, int[] list) {
        int max = 1;

        for (int i = 1; i <= n; i++) {
            d[i] = 1;
            for (int j = 1; j < i; j++) {
                if (list[i] > list[j] && d[i] <= d[j])
                    d[i] = d[j] + 1;
            }
            if (max < d[i])
                max = d[i];
        }

        return max;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        int N;
        int[] A;
        int result = 0;

        N = Integer.parseInt(br.readLine());
        A = new int[MAX + 1];
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++)
            A[i] = Integer.parseInt(st.nextToken());

        result = lis_dp(N, A);

        bw.write(String.valueOf(result));
        bw.flush();
        bw.close();
    }
}

File: Dynamic Programming/11053. 가장 긴 증가하는 부분 수열/11053.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>

#define MAX 1000

int d[MAX + 1] = { 0, };

int lis_dp(int n, int list[]) {
	int max = 1;

	for (int i = 1; i <= n; i++) {
		d[i] = 1;
		for (int j = 1; j < i; j++) {
			if (list[i] > list[j] && d[i] <= d[j])
				d[i] = d[j] + 1;
		}
		if (max < d[i])
			max = d[i];
	}

	return max;
}

int main() {
	int N;
	int A[MAX + 1];
	int result;

	scanf("%d", &N);
	for (int i = 1; i <= N; i++)
		scanf("%d", &A[i]);

	result = lis_dp(N, A);

	printf("%d", result);

	return 0;
}